lib/functions.php

```diff
@@ -223,6 +223,47 @@ function timeout($promise, int $timeout): Promise
         return $deferred->promise();
     }
 
+    /**
+     * Creates an artificial timeout for any `Promise`.
+     *
+     * If the promise is resolved before the timeout expires, the result is returned
+     *
+     * If the timeout expires before the promise is resolved, a default value is returned
+     *
+     * @param \Amp\Promise|\React\Promise\PromiseInterface $promise Promise to which the timeout is applied.
+     * @param int                                          $timeout Timeout in milliseconds.
+     * @param mixed                                        $default
+     *
+     * @return \Amp\Promise
+     *
+     * @throws \TypeError If $promise is not an instance of \Amp\Promise or \React\Promise\PromiseInterface.
+     */
+    function timeoutWithDefault($promise, int $timeout, $default = null): Promise
+    {
+        if (!$promise instanceof Promise) {
+            if ($promise instanceof ReactPromise) {
+                $promise = adapt($promise);
+            } else {
+                throw createTypeError([Promise::class, ReactPromise::class], $promise);
+            }
+        }
+
+        $promise = timeout($promise, $timeout);
+
+        $deferred = new Deferred();
+        $newPromise = $deferred->promise();
+
+        $promise->onResolve(function ($exception, $value) use ($deferred, $default) {
+            if ($exception) {
+                $deferred->resolve($default);
```

Comment:
An exception should probably be thrown here instead of resulting in the default value.

Reply:
You're right, give me a minute.

Reply:
I wouldn't make use of the timeout function, but rather adapt the code of it to suit this function.

Reply:
No wait, that's the whole purpose of the default value. You timeout with a default and when you'll receive a timeout exception, you'll get the default value.

Reply:
You'll get the default on a timeout, but not if the promise fails?